litedram/phy/ddr5/commands.py

```diff
@@ -0,0 +1,227 @@
+        ]
+
+        dfi_cmd = Signal(3)
+        self.comb += dfi_cmd.eq(Cat(~dfi_phase.we_n, ~dfi_phase.ras_n, ~dfi_phase.cas_n)),
+        _cmd = {  # cas, ras, we
+            "NOP": 0b000,
+            "ACT": 0b010,
+            "RD":  0b100,
+            "WR":  0b101,
+            "PRE": 0b011,
+            "REF": 0b110,
+            "ZQC": 0b001,
+            "MRS": 0b111,
+        }
+
+        def cmds(cmd2, valid=1):
+            return self.cmd1.set("DESELECT") + self.cmd2.set(cmd2) + [self.valid.eq(valid)]
+
+        self.comb += If(dfi_phase.cs_n == 0,  # require dfi.cs_n
+            Case(dfi_cmd, {
+                _cmd["ACT"]: cmds("ACTIVATE"),
+                _cmd["RD"]:  cmds("READ"),
+                _cmd["WR"]:  cmds("WRITE"),
+                _cmd["PRE"]: cmds("PRECHARGE ALL"),
+                _cmd["REF"]: cmds("REFRESH ALL"),
+                # Use bank address to select command type
+                _cmd["ZQC"]: Case(dfi_phase.bank, {
+                    SpecialCmd.MPC: cmds("MPC"),
+                    SpecialCmd.MRR: cmds("MRR"),
+                    "default": cmds("DESELECT", valid=0),
+                }),
+                _cmd["MRS"]: cmds("MRW"),
+                "default": cmds("DESELECT", valid=0),
+            })
+        )
+
+
+class Command(Module):
+    """DDR5 command decoder
+
+    Decodes a command from single DFI phase into DDR5 "small command"
+    consisting of 2 CS values and 2 CA[5:0] values.
+
+    DDR5 "small commands" are transmited over 2 clock cycles. In the first
+    cycle CS is driven high and in the second cycle it stays low. In each
+    of the cycles the bits on CA[5:0] are latched and interpreted differently.
+    This module translates a DFI command into the values of CS/CA that shall
+    be transmitted over 2 DRAM clock cycles.
+
+    Attributes
+    ----------
+    dfi : Record(dfi.phase_description), in
+        Input from single DFI phase.
+    cs : Signal(2), out
+        CS values over 2 subsequent DRAM SDR clock cycles.
+    ca : Array(2, Signal(14)), out
+        CA[13:0] values over 2 subsequent DRAM SDR clock cycles.
+    """
+
+    # String description of 1st and 2nd edge of each command, later parsed to
+    # construct the value. CS is assumed to be H for 1st edge and L for 2nd edge.
+    # TODO: Fix currently fake truth table to be consistent with the JEDEC specs for DDR5
+    # Here, in contrary to LPDDR4's table, first element of array is related to CS low,
+    # the second to CS high. Something will have to be changed, because other parts of code
+    # rely on order from LPDDR4, but there are commands that have only row related to low.
+    # At this moment I keep it as it is now.
+    TRUTH_TABLE = {
+        "ACTIVATE":      ["L L L R0 R1 R2 R3 BA0 BA1 BG0 BG1 BG2 CID0 CID1 CID2",
+                          "R4 R5 R6 R7 R8 R9 R10 R11 R12 R13 R14 R15 R16, CID3/R17"],
+        "READ":          ["H L H H H BL*=L BA0 BA1 BG0 BG1 BG2 CID0 CID1 CID2",
```

Comment:
I'd keep just `BL` instead of `BL*=L`. It is too verbose otherwise.

Reply:
Done